### ISSUE TITLE
Setting nested properties not working correctly

### ISSUE THREAD

**Issue by paglias** [Automattic/mongoose#7748]
This is actually extremely similar to #7707. I'm opening a new issue as requested by @vkarpov15 .

Since 5.4.19 our build has been failing because it looks like nested properties are not being set correctly (or even set to undefined). 

In our own case this is happening with Number fields that are set to undefined and then when mongoose tries to cast them to numbers, they become `NaN` (`Number(undefined) -> NaN`) https://travis-ci.org/HabitRPG/habitica/jobs/524634104

I can followup with a different repro script if needed but this code posted in #7707 by @Msordet works in reproducing the issue

```
const mongoose = require("mongoose");
const Schema = mongoose.Schema;

// Fix all deprecation warnings: https://mongoosejs.com/docs/deprecations.html
mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);

const geojsonSchema = new Schema(
  {
    type: { type: String, default: "Feature" },
    geometry: {
      type: {
        type: String,
        enum: [
          "Point",
          "MultiPoint",
          "LineString",
          "MultiLineString",
          "Polygon",
          "MultiPolygon",
          "GeometryCollection"
        ],
        required: true
      },
      coordinates: { type: [] }
    },
    properties: { type: Object }
  },
  {
    _id: false,
    timestamps: {
      createdAt: true
    }
  }
).index({ geometry: "2dsphere" });

const userSchema = new Schema({
  position: geojsonSchema
});

const GeoJson = mongoose.model("GeoJson", geojsonSchema);
const User = mongoose.model("User", userSchema);

mongoose.connect("mongodb://localhost:27017/test-nested").then(async () => {
  const position = new GeoJson({
    geometry: {
      type: "Point",
      coordinates: [1.11111, 2.22222]
    },
    properties: {
      a: "b"
    }
  });

  const newUser = new User({
    position
  });
  try {
    await newUser.save();
  } catch (e) {
    console.log("Fail at user creation");
    console.error(e);
  }

  const editUser = await User.findById(newUser._id);

  editUser.position = position;

  try {
    await position.validate();
    await editUser.validate();
  } catch (e) {
    console.log("Fail at user validation");
    console.error(e);
  }
});
```

**Comment by Fonger** [comment]
After investigation, I find out that this is caused by this line https://github.com/Automattic/mongoose/commit/6c66ba875cf38ff865198505c1b3bf4db18e4fd7#diff-8966d6fe57bc9430faa5e59c750deecfR41
, but I'm not fully understand the code here so I afraid to change it.

A simpler repro script, basically, this can be reproduced you have directly inlined schema under a nestedSchema.


```js
const mongoose = require("./lib");
const Schema = mongoose.Schema;

mongoose.set('debug', true);
mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);

const nestedSchema = new Schema({
  name: { type: String, required: true },
  nestedWithoutSchema: {
    deep: { type: String, required: true }
  }
})

const userSchema = new Schema({
  nestedWithSchema: nestedSchema // if you inline all schema here, everything works fine.
});

const User = mongoose.model("User", userSchema);

mongoose.connect("mongodb://localhost:27017/test-nested").then(async () => {
  const newUser = new User({
    nestedWithSchema: {
      name: 'Fonger',
      nestedWithoutSchema: {
        deep: 'very deep!'
      }
    }
  });

  await newUser.save();

  const editUser = await User.findById(newUser._id);
  editUser.nestedWithSchema = { name: 'Fonger2', nestedWithoutSchema: { deep: 'deeper' } }
  // nestedWithoutSchema was removed here so validation fails.

  try {
    await editUser.validate();
  } catch (e) {
    console.log("Fail at user validation");
    console.error(e);
  }
});
```

**Comment by vkarpov15** [comment]
Sorry for the trouble @paglias , we did some work to make sure single nested setters could modify other properties in single nested subdocs and it has caused a few issues. I believe I fixed this issue, will double check @Fonger 's repro script later.

**Comment by paglias** [comment]
@vkarpov15 thanks! I can confirm that this commit actually fixes the issue

**Comment by vkarpov15** [comment]
Confirmed, fixed by 5aa79c2

**Comment by vkarpov15** [comment]
Thanks for reporting this issue @paglias :+1:
